pkg/restic/exec_commands.go

```diff
@@ -53,3 +62,91 @@ func GetSnapshotID(repoIdentifier, passwordFile string, tags map[string]string,
 
 	return snapshots[0].ShortID, nil
 }
+
+// RunBackup runs a `restic backup` command and watches the output to provide
+// progress updates to the caller.
+func RunBackup(backupCmd *Command, log logrus.FieldLogger, updateFunc func(velerov1api.PodVolumeOperationProgress)) (string, string, error) {
+	// buffers for copying command stdout/err output into
+	stdoutBuf := new(bytes.Buffer)
+	stdoutStatBuf := new(bytes.Buffer)
+	stderrBuf := new(bytes.Buffer)
+	// stdoutWriter is used to copy the command's stdout to both stdoutStatBuf and
+	// stdoutBuf, the former buffer is used to scan through the output and the
+	// latter is used to return the command's output as a string
+	stdoutWriter := io.MultiWriter(stdoutStatBuf, stdoutBuf)
+
+	// create a channel to signal when to end the goroutine scanning for progress
+	// updates
+	quit := make(chan struct{})
+	var wg sync.WaitGroup
+
+	cmd := backupCmd.Cmd()
+	stdout, err := cmd.StdoutPipe()
+	if err != nil {
+		return "", "", err
+	}
+	stderr, err := cmd.StdoutPipe()
+	if err != nil {
+		return "", "", err
+	}
+
+	cmd.Start()
+
+	// copy command's stdout/err to buffers
+	wg.Add(1)
+	go func() {
```

Comment:
are these 2 goroutines for copying stdout/stderr needed, or can we just set `cmd.Stdout = stdoutWriter` and similarly for stderr?

Reply:
Good point, let me check - I assumed it wouldn't work because I tried to use `cmd.StdoutPipe()` directly to scan through but with that `scanner.Scan()` would continuously stream until the command was finished, so the buffer is needed to read up to the current point. Let's see if I can set `cmd.Stdout = stdoutWriter` and it behaves the same way.

Reply:
Looks like that works, thanks for the tip!